### PR TITLE
ntile with only one argument in dplyr expressions

### PR DESCRIPTION
**Pull request by romainfrancois**
```
> df <- tibble(x=1:42)
> mutate( df, nt = ntile(9))
# A tibble: 42 x 2
       x    nt
   <int> <int>
 1     1     1
 2     2     1
 3     3     1
 4     4     1
 5     5     1
 6     6     2
 7     7     2
 8     8     2
 9     9     2
10    10     2
# ... with 32 more rows
```

on the example from #3418 :

```r
> library(tidyr)
>  mtcars %>%
+   as_tibble() %>% 
+   mutate(ntile = ntile(7)) %>%
+   nest(-ntile)
# A tibble: 7 x 2
  ntile data             
  <int> <list>           
1     1 <tibble [5 × 11]>
2     2 <tibble [5 × 11]>
3     3 <tibble [4 × 11]>
4     4 <tibble [5 × 11]>
5     5 <tibble [4 × 11]>
6     6 <tibble [5 × 11]>
7     7 <tibble [4 × 11]>
```

I have not touched the non hybrid version, so at the moment these are confusing

```
> ntile(4)
Error in ntile(4) : argument "n" is missing, with no default
> ntile(n=4)
Error in ntile(n = 4) : argument "x" is missing, with no default
```

### CODE REVIEW

**Comment by krlmlr** [comment]
I like Hadley's suggestion to

> set `row_number()` as the first argument, so you can just write `ntile(n = 7)`

per https://github.com/tidyverse/dplyr/issues/3418#issuecomment-372820382.

**Comment by romainfrancois** [comment]
That’s what i’ve done conceptually. But knowing in advance that this is row_number gives me an advantage. This is what hybrid is all about.

**Comment by krlmlr** [comment]
I think hybrid should use special behavior only with `ntile(n = 9)` and not with `ntile(9)`, and the standard version should be declared as `ntile(x = row_number(), n)` . Maybe it's easier to define a new rank function `ntile_rows(n)` ?

**Comment by romainfrancois** [comment]
With the last commit you'd get: 

```
> mutate( df, nt = ntile(9))
 Show Traceback
 
 Rerun with Debug
 Error in mutate_impl(.data, dots) : 
  Evaluation error: argument "n" is missing, with no default. 
> mutate( df, nt = ntile(n=9))
# A tibble: 42 x 2
       x    nt
   <int> <int>
 1     1     1
 2     2     1
 3     3     1
 4     4     1
 5     5     1
 6     6     2
 7     7     2
 8     8     2
 9     9     2
10    10     2
# ... with 32 more rows
```

I'm not convinced

File: src/hybrid_window.cpp [hunk 1]
```diff
@@ -166,3 +182,47 @@ void install_window_handlers(HybridHandlerMap& handlers) {
...
+  int ng = gdf.ngroups();
+
+  GroupedDataFrame::group_iterator git = gdf.group_begin();
+  IntegerVector out = no_init(n);
+  for (int i = 0; i < ng; i++, ++git) {
+    const SlicingIndex& index = *git;
+    int m = index.size();
+
+    for (int j = m - 1; j >= 0; j--) {
+      out[ index[j] ] = (int)floor((ntiles * j) / m) + 1;
```
Comment (krlmlr) [inline src/hybrid_window.cpp]: Would you like to extract this arithmetic operation to a function?

File: src/hybrid_window.cpp [hunk 2]
```diff
@@ -81,30 +81,46 @@ Result* ntile(const RObject& data, const int number_tiles, const bool ascending)
...
 
-  if (TYPEOF(data) == SYMSXP) {
-    SymbolString name = SymbolString(Symbol(data));
-    if (subsets.has_non_summary_variable(name)) data = subsets.get_variable(name);
-    else return 0;
+    return new Ntile_1(number_tiles);
   }
 
-  if (subsets.nrows() != Rf_length(data)) return 0;
+  if (nargs == 2) {
```
Comment (krlmlr) [inline src/hybrid_window.cpp]: `else`, just to indicate the relation to the preceding `if`?

**Comment by krlmlr** [comment]
I think consistency with the R version of the function is very important. So far the only documented inconsistencies are `row_number()` and `n()`.

We might need to extend/improve the documentation.

**Comment by romainfrancois** [comment]
Merging this now so that I can tackle #3526

**Pull request merged**